consensus/exit_probability_randomizer/exit_prob_map.c

```diff
@@ -0,0 +1,233 @@
+             ep_validator, tip_entry->hash, &has_valid_tail)) != RC_OK) {
+      log_error(EXIT_PROB_MAP_LOGGER_ID,
+                "Tail transaction validation failed: %" PRIu64 "\n", ret);
+      goto done;
+    }
+    if (!has_valid_tail) {
+      if (!hash_to_indexed_hash_set_map_find(
+              &cw_result->tx_to_approvers, tip_entry->hash, &approvers_entry)) {
+        goto done;
+      }
+      hash243_set_remove(&approvers_entry->approvers, tip_entry->hash);
+    }
+  }
+
+done:
+  hash243_set_free(&tips);
+  return ret;
+}
+
+/*
+ * Public functions
+ */
+
+void iota_consensus_exit_prob_map_init(ep_randomizer_t *const randomizer) {
+  randomizer->base.vtable = exit_prob_map_vtable;
+  ep_prob_map_randomizer_t *prob_randomizer =
+      (ep_prob_map_randomizer_t *const)randomizer;
+  prob_randomizer->exit_probs = NULL;
+  prob_randomizer->transition_probs = NULL;
+}
+
+retcode_t iota_consensus_exit_prob_map_randomize(
+    ep_randomizer_t const *const randomizer,
+    exit_prob_transaction_validator_t *const ep_validator,
+    cw_calc_result *const cw_result, flex_trit_t const *const ep,
+    flex_trit_t *tip) {
+  retcode_t ret;
+  ep_prob_map_randomizer_t *prob_randomizer =
+      (ep_prob_map_randomizer_t *const)randomizer;
+
+  double rand_weight;
+
+  if (prob_randomizer->exit_probs == NULL) {
+    if ((ret = iota_consensus_exit_prob_map_calculate_probs(
+             randomizer, ep_validator, cw_result, ep,
+             &prob_randomizer->exit_probs,
+             &prob_randomizer->transition_probs)) != RC_OK) {
+      return ret;
+    }
+  }
+
+  hash243_set_t tips = NULL;
+  iota_consensus_exit_prob_map_eps_extract_tips(cw_result, &tips);
+  // Randomizes uniformly a value between 0 to 1
+  rand_weight = rand_handle_probability();
+  hash243_set_entry_t *tip_entry = NULL;
+  hash243_set_entry_t *tip_tmp_entry = NULL;
+  hash_to_double_map_entry_t *curr_ep = NULL;
+
+  HASH_ITER(hh, tips, tip_entry, tip_tmp_entry) {
+    hash_to_double_map_find(&prob_randomizer->exit_probs, tip_entry->hash,
+                            &curr_ep);
+    rand_weight -= curr_ep->value;
+    if (rand_weight <= 0) {
+      memcpy(tip, tip_entry->hash, FLEX_TRIT_SIZE_243);
+      break;
+    }
+  }
+  hash243_set_free(&tips);
+  return RC_OK;
+}
+
+retcode_t iota_consensus_exit_prob_map_calculate_probs(
+    ep_randomizer_t const *const exit_probability_randomizer,
+    exit_prob_transaction_validator_t *const ep_validator,
+    cw_calc_result *const cw_result, flex_trit_t const *const ep,
+    hash_to_double_map_t *const hash_to_exit_probs,
+    hash_to_double_map_t *const hash_to_trans_probs) {
+  retcode_t ret;
+  bool ep_is_valid = false;
+  hash243_queue_t queue = NULL;
+  hash_to_indexed_hash_set_entry_t *aps = NULL;
+  hash_to_double_map_entry_t *curr_tx_entry = NULL;
+  hash243_set_entry_t *approver_entry = NULL;
+  hash243_set_entry_t *tmp_entry = NULL;
+  size_t num_approvers;
```

Comment:
init to default values?

Reply:
`num_approvers` does not have a default value, and both vars won't be read before being set